### DISCUSSION TITLE
Looking for best option

### DISCUSSION THREAD

**Discussion by rexton12** [mirumee/ariadne#1086] [General]
What are the benefits of using Ariadne over other GraphQL libraries in Python, such as Graphene or Strawberry?

How does Ariadne handle error handling and validation of input data in GraphQL queries and mutations?

How does Ariadne handle authorization and authentication in GraphQL APIs?

How does Ariadne handle performance and scalability in GraphQL APIs, especially with larger and more complex queries and data sets?

How does Ariadne handle versioning and backward compatibility of GraphQL APIs, especially in larger and more complex applications?

How does Ariadne compare to other GraphQL libraries in terms of ease of use and learning curve for developers who are new to GraphQL?

What are some best practices for using Ariadne in production applications, including testing, debugging, and deployment strategies?

How can the Ariadne community continue to improve and evolve the library, and what features or functionality would be most valuable to developers working with GraphQL in Python?

**Comment by rafalp** [comment]
Hello!

Ariadne, Strawberry and Graphene have a lot in common, because all three of those libraries are just _wrappers_ for [graphql core](https://github.com/graphql-python/graphql-core/releases) library that does all the GraphQL logic:

- error handling and validation of input data in GraphQL queries and mutations
- performance and scalability in GraphQL APIs, especially with larger and more complex queries and data sets
- input deserialization and validation
- error handling

Only differences are in Python logic that extracts data from HTTP payload and puts to `graphql.execute()` function from graphql core, extras available out of the box, and what approach they use for GraphQL schema definition. Ariadne is schema-first which we believe is easier to learn because of existing amount of GraphQL learning resources that also use schema first approach to teaching GraphQL.

Unlike, say, REST API which is more like a set of loose rules to follow, GraphQL has [strict specification](https://spec.graphql.org/October2021/) which defines that GraphQL servers should do, and how.

> Ariadne handle authorization and authentication in GraphQL APIs

This is out of scope for GraphQL APIs. Our recommendation is that you implement it outside of GraphQL, in your HTTP framework of choice (be it FastAPI or Django) and only expose information about the user in GraphQL `context` data.

> handle versioning and backward compatibility of GraphQL APIs, especially in larger and more complex applications

GraphQL specification doesn't provide any opinion about evolving your GraphQL schema besides of inclusion of `@deprecated` directive that can be used to annotate schema elements. It's up to developers to decide which schema versioning approach they'll employ in their project.

> What are some best practices for using Ariadne in production applications, including testing, debugging, and deployment strategies?

This is a VERY general question. Could you be more specific? We generally support all Python web frameworks and both WSGI and ASGI deployments. We also make no assumptions on your testing setup, but recommended approach for testing GraphQL APIs is using your frameworks test client to run mock HTTP requests with GraphQL payloads against your GraphQL API.

> How can the Ariadne community continue to improve and evolve the library, and what features or functionality would be most valuable to developers working with GraphQL in Python?

I am not sure what you are asking for here? If you are looking to contribute there are few issues on Ariadne's GitHub that have "help wanted" label. :)
